crates/goose/src/providers/ollama.rs

```diff
@@ -0,0 +1,378 @@
+        let usage = Self::get_usage(&response)?;
+
+        Ok((message, usage))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use serde_json::json;
+    use wiremock::matchers::{method, path};
+    use wiremock::{Mock, MockServer, ResponseTemplate};
+
+    // Helper function to create a provider with mock server
+    async fn setup_mock_provider(response_body: Value) -> Result<OllamaProvider> {
+        let mock_server = MockServer::start().await;
+        
+        // Mock the health check endpoint
+        Mock::given(method("GET"))
+            .and(path("/"))
+            .respond_with(ResponseTemplate::new(200))
+            .mount(&mock_server)
+            .await;
+
+        // Mock the completions endpoint
+        Mock::given(method("POST"))
+            .and(path("/v1/chat/completions"))
+            .respond_with(ResponseTemplate::new(200).set_body_json(response_body))
+            .mount(&mock_server)
+            .await;
+
+        let config = OllamaProviderConfig {
+            host: mock_server.uri(),
+        };
+
+        OllamaProvider::new(config).await
+    }
+
+    // Helper function to create a provider for integration tests
+    async fn setup_integration_provider() -> Result<OllamaProvider> {
+        let config = OllamaProviderConfig {
+            host: std::env::var("OLLAMA_HOST")
+                .unwrap_or_else(|_| String::from("http://localhost:11434")),
+        };
+        OllamaProvider::new(config).await
+    }
+
+    // Helper function for completion tests
+    async fn test_completion(provider: &OllamaProvider) -> Result<(Message, Usage)> {
+        let messages = vec![Message::user("Hello?")?];
+        provider
+            .complete(
+                OLLAMA_MODEL,
+                "You are a helpful assistant.",
+                &messages,
+                &[],
+                None,
+                None,
+            )
+            .await
+    }
+
+    // Helper function for tool tests
+    async fn test_tools(provider: &OllamaProvider) -> Result<(Message, Usage)> {
+        let messages = vec![Message::user("Can you read the test.txt file?")?];
+        let tool = Tool::new(
+            "read_file",
+            "Read the content of a file",
+            json!({
+                "type": "object",
+                "properties": {
+                    "filename": {
+                        "type": "string",
+                        "description": "The name of the file to read"
+                    }
+                },
+                "required": ["filename"]
+            }),
+        );
+
+        provider
+            .complete(
+                OLLAMA_MODEL,
+                "You are a helpful assistant.",
+                &messages,
+                &[tool],
+                None,
+                None,
+            )
+            .await
+    }
+
+    #[tokio::test]
+    async fn test_complete_basic() -> Result<()> {
```

Comment:
in the python side we started with hand crafted assertion style. one problem with this is that things can be easily missed from what the real providers return, and also it is errorprone to adjust say when ollama 0.5.0 is released. So, we went to VCR tests which allowed us to capture real requests

Ack in rust the easiest to find VCR port has been abandoned a few years ago, but maybe there's another out there somewhere? https://docs.rs/surf-vcr/latest/surf_vcr/

Reply:
Updated to use httpmock, please take a look

Reply:
actually. I undid it, open-ai and databricks do not use this so right now it is consistent with those tests, I was also hitting some issues with getting the recorder set up here, I do think we should think more about your comment  but in a separate PR so will handle it separately.